components/Home/Bonding/requests.jsx

```diff
@@ -402,51 +403,55 @@ const getLpPriceWithProjectedChange = (list) => list.map((record) => {
   };
 });
 
-/**
- *
- */
+const MAX_CONCURRENT_REQUESTS = 5;
+
+/* eslint-disable */
 const getProductDetailsFromIds = async ({ productIdList }) => {
   const contract = getDepositoryContract();
+  const discount = await getLastIDFRequest();
 
-  const allListPromise = [];
-  for (let i = 0; i < productIdList.length; i += 1) {
-    const id = productIdList[i];
-    const allListResult = contract.methods.mapBondProducts(id).call();
-    allListPromise.push(allListResult);
-  }
+  const createEventList = await getCreateProductEvents();
+  const closedEventList = await getCloseProductEvents();
 
-  // discount factor is same for all the products
-  const discount = await getLastIDFRequest();
+  const productDetails = [];
 
-  const response = await Promise.all(allListPromise);
-  const productList = response.map((product, index) => ({
-    ...product,
-    discount,
-    id: productIdList[index],
-  }));
+  // Split productIdList into chunks to limit concurrent requests
+  for (let i = 0; i < productIdList.length; i += MAX_CONCURRENT_REQUESTS) {
+    const slicedIds = productIdList.slice(i, i + MAX_CONCURRENT_REQUESTS);
```

Comment:
This is likely a great opportunity to use multicall instead of looping and issuing multiple requests.

Multicall contracts allow you to batch requests into one request on a specific chain and returns an array of responses.

@Tanya-atatakai see: https://www.multicall3.com/

Ethers or Viem have simple packages for this; with web3 which we're using in this context, we'll have to make a custom implementation unless you can find a package for it.

Reply:
This would contribute to much faster loading times, the current implementation is much too slow :(

Reply:
Here's the usage for Viem: https://viem.sh/docs/contract/multicall.html#usage
And the Github for ethers-multicall: https://github.com/cavanmflynn/ethers-multicall/tree/master

I appreciate they're not web3 packages, but can give an idea of the inputs and expected responses.

Reply:
thanks for this advice @truemiller , works much better with the multicall implementation 🙏